kernel/kernel-api/src/main/java/io/delta/kernel/internal/SnapshotImpl.java

```diff
@@ -263,29 +263,30 @@ public void publish(Engine engine) throws PublishFailedException {
 
   @Override
   public void writeChecksum(Engine engine, Snapshot.ChecksumWriteMode mode) throws IOException {
-    final Snapshot.ChecksumWriteMode actual = getStatistics().getChecksumWriteMode();
+    final Optional<Snapshot.ChecksumWriteMode> actualOpt = getStatistics().getChecksumWriteMode();
 
-    if (actual != mode) {
-      logger.warn("Requested {} mode but actual mode is {}.", mode, actual);
+    if (actualOpt.isEmpty()) {
+      logger.warn("Not writing checksum: checksum file already exists at version {}", version);
+      return;
     }
 
+    final Snapshot.ChecksumWriteMode actual = actualOpt.get();
+
     switch (mode) {
-      case NONE:
-        logger.info("Skipping writing checksum file: input mode was NONE");
-        return;
       case SIMPLE:
-        final CRCInfo crcInfo =
-            logReplay
-                .getCrcInfoAtSnapshotVersion()
-                .orElseThrow(
-                    () ->
-                        new IllegalStateException(
-                            "Cannot write checksum in mode SIMPLE: info not available"));
+        if (actual == ChecksumWriteMode.FULL) {
+          throw new IllegalStateException(
+              "Cannot write checksum in SIMPLE mode: FULL mode required");
+        }
 
+        final CRCInfo crcInfo = logReplay.getCrcInfoAtSnapshotVersion().get();
         logger.info("Executing checksum write in SIMPLE mode");
         new ChecksumWriter(logPath).writeCheckSum(engine, crcInfo);
         return;
       case FULL:
+        if (actual == ChecksumWriteMode.SIMPLE) {
+          logger.warn("Requested checksum write in FULL mode, but SIMPLE mode is available");
+        }
         logger.info("Executing checksum write in FULL mode");
         ChecksumUtils.computeStateAndWriteChecksum(engine, getLogSegment());
```

Comment:
Can't we just do `new ChecksumWriter(logPath).writeCheckSum(engine, crcInfo)` here when `actual == ChecksumWriteMode.SIMPLE`?

Reply:
Not sure if computeStateAndWriteChecksum shortcuts that when possible?

Reply:
Maybe we should switch on the actual mode instead?